src/connectors/userService.ts

```diff
@@ -2169,6 +2178,250 @@ export class UserService extends BaseService {
     return updatedUser
   }
 
+  /*********************************
+   *                               *
+   *        Social Login           *
+   *                               *
+   *********************************/
+
+  public getOrCreateUserBySocialAccount = async ({
+    type,
+    providerAccountId,
+    userName,
+    email,
+    emailVerified,
+  }: SocialAccount & { emailVerified?: boolean }) => {
+    // check if social account exists, if true, return user directly
+    const socialAcount = await this.getSocialAccount({
+      type,
+      providerAccountId,
+      userName,
+    })
+    let user
+    if (socialAcount) {
+      user = await this.loadById(socialAcount.userId)
+      if (!user.emailVerified && emailVerified) {
+        return this.baseUpdate(user.id, { emailVerified })
+      }
+      return user
+    }
+
+    // social account not exists, create social account and user if not exists
+    if (email) {
+      user = await this.findByEmail(email)
+    }
+    const trx = await this.knex.transaction()
+    let isCreated = false
+    try {
+      if (!user) {
+        user = await this.create({ email, emailVerified }, trx)
+        isCreated = true
+      } else {
+        if (!user.emailVerified && emailVerified) {
+          user = await this.baseUpdate(
+            user.id,
+            { emailVerified },
+            undefined,
+            trx
+          )
+        }
+      }
+      await this.createSocialAccount(
+        { userId: user.id, type, providerAccountId, userName, email },
+        trx
+      )
+      await trx.commit()
+    } catch (error) {
+      await trx.rollback()
+      throw error
+    }
+    if (isCreated) {
+      await this.postRegister(user)
+    }
+    return user
+  }
+
+  private getSocialAccount = async ({
+    type,
+    providerAccountId,
+  }: SocialAccount) => {
+    return await this.knex('social_account')
+      .select()
+      .where({ type, providerAccountId })
+      .first()
+  }
+
+  private createSocialAccount = async (
+    { userId, type, providerAccountId, userName, email }: SocialAccount,
+    trx?: Knex.Transaction
+  ) => {
+    const query = this.knex('social_account')
+      .insert({ userId, type, providerAccountId, userName, email })
+      .returning('*')
+
+    if (trx) {
+      query.transacting(trx)
+    }
+
+    return query
+  }
+
+  /**
+   * Fetch the Twitter user info using Twitter v2 API.
+   * @see {@link https://developer.twitter.com/en/docs/authentication/oauth-2-0/user-access-token}
+   */
+  public fetchTwitterUserInfo = async (
+    authorizationCode: string,
+    codeVerifier: string
+  ) => {
+    const accessToken = await this.exchangeTwitterAccessToken(
+      authorizationCode,
+      codeVerifier
+    )
+    return await this.fetchTwitterUserInfoByAccessToken(accessToken)
```

Comment:
`await` in the last return is always unnecessary

Reply:
weird, npm run format && npm run lint:fix should fix this